docs(aio): Clearing array with [] (#20369)

ShaviPathania opened the pull request:
Clearing array with setting length to 0 replaced with [] for being short
and marginally efficient. For reference: [] is turned into a sequence of
around 15 machine instructions on x64 (if bump pointer allocation succeeds),
whereas a.length=0 is a C++ runtime function call, which requires 10-100x as
many instructions.
Benedikt Meurer

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #20369


## What is the new behavior?
Arrays are being cleared using [] instead of setting it's length to 0

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

ShaviPathania committed:
docs(aio): Clearing array with [] (#20369)

Clearing array with setting length to 0 replaced with [] for being short
and marginally efficient. For reference: [] is turned into a sequence of
around 15 machine instructions on x64 (if bump pointer allocation succeeds),
whereas a.length=0 is a C++ runtime function call, which requires 10-100x as
many instructions.
Benedikt Meurer

shavipathania commented:
I signed it!

gkalpak commented on aio/content/examples/lifecycle-hooks/src/app/logger.service.ts:
```diff
@@ -18,7 +18,7 @@ export class LoggerService {
   }
 
-  clear() { this.logs.length = 0; }
+  clear() { this.logs = []; }
```
This will not work as it is, because `SpyComponent` holds a reference to the previous array.
(This is what I was referring to in https://github.com/angular/angular/issues/20369#issuecomment-343892856.)

Other examples are fine, but the `lifecycle-hooks` guide needs further modifications if you want to move to the ` = []` style.

[1]: https://github.com/shavipathania/angular/blob/33740a69db38d44f514d09e616bd0340a3cc30e3/aio/content/examples/lifecycle-hooks/src/app/spy.component.ts#L21

shavipathania replied:
Thanks for checking my pull request...

Sorry, but I didn't get what issue could arise...

Logger reset using [] - [Plunker](https://embed.plnkr.co/PK01rFNZrmtdzFbAR9iC/)

Logger reset using .length = 0 - [Plunker](https://angular.io/generated/live-examples/lifecycle-hooks/eplnkr.html)

They both works same... on spy component.... and there are no errors in console....

(I think the issue can arise when initially log service is initialised with some extended version of Core JS Array. But in that case it's responsibility of developer to reset it with that extended version.)

gkalpak replied:
It maky work because the `clear()` method is never used (afaict from a quick look). If it were used, `SpyComponent` would not work as expected.
If it is indeed never used, maybe remove the method altogether.

shavipathania replied:
I suggest if instead of removing clear() method from LogService, we should do following

spy.component.ts
```typescript 
// making logger public so can be accessed in template
constructor(public logger: LoggerService) {
}
```


spy.component.html
```xml
<!-- directly using logger.logs instead of using spyLog -->
<div *ngFor="let msg of logger.logs">{{msg}}</div>
```

Doing so will be more aligned with guidelines as suggested by you it's not good idea to keep reference to Service properties.

gkalpak replied:
I don't mind, but why not remove a method that is not used?

shavipathania replied:
because it's being used in `peek-a-boo-parent.component.ts`
```typescript
  toggleChild() {
    this.hasChild = !this.hasChild;
    if (this.hasChild) {
      this.heroName = 'Windstorm';
      this.logger.clear(); // clear log on create
    }
    this.logger.tick();
  }

```
Also, I have made changes in spy component. Please have a [look](https://github.com/angular/angular/pull/20395/commits/29c7358354cfa66fa5b2d046fb52b7f44590cde9)

gkalpak replied:
> because it's being used in `peek-a-boo-parent.component.ts`

Oh, nice. You can see how ` = []` breaks there, then :grin:

ShaviPathania committed:
docs(aio): Removing reference to LoggerService property (#20369)

The reference removed so that calling LoggerService clear() method
behaves as intended in SpyComponent.

gkalpak commented on aio/content/examples/lifecycle-hooks/src/app/after-content.component.ts:
```diff
@@ -108,7 +108,7 @@ export class AfterContentParentComponent {
 
   reset() {
-    this.logs.length = 0;
+    this.logs = [];
```
Previously, this line emptied the `logger.logs` array. With this change, adding logs to `logger.logs` will have no effect on `this.logs` any more. The same applies to other components.

You could use `this.logger.clear()` instead? But then you have to make appropriate chnges to all components that use `this.logs = logger.logs` and keep a reference to the `logger.logs` array.

shavipathania replied:
Thanks, for pointing out.... :+1: 

I will be making necessary changes tonight... :-)

shavipathania replied:
Made necessary changes, things are working fine... 

for your quick reference...
Original Example: [Plunker](https://angular.io/generated/live-examples/lifecycle-hooks/eplnkr.html)
Only new Changes (logger service + two newly modified files): [Plunker](https://embed.plnkr.co/HHUjVazDSokwBJIKSXpJ/)

have a look at new [commit](https://github.com/angular/angular/pull/20395/commits/8dfd7f17e4603f1027573ec63bf5f890a276612e)

ShaviPathania committed:
docs(aio): Clearing array with [] (#20369)

Convert remaining references to directly use LoggerServices logs.

mary-poppins commented:
You can preview 33740a6 at https://pr20395-33740a6.ngbuilds.io/.
You can preview 29c7358 at https://pr20395-29c7358.ngbuilds.io/.
You can preview 8dfd7f1 at https://pr20395-8dfd7f1.ngbuilds.io/.
